kwasm-lemmas.md

```diff
@@ -20,13 +20,157 @@ Basic logic
 Basic arithmetic
 ----------------
 
+// TODO: Upstream the lemmas in this section into K.
+
+### Modular Arithmetic
+
+Z3 is slow and unreliable in reasoning about modular arithmetic.
+Therefore we want to make structural simplifications wherever possible.
+
+#### Rules for Expressions With Only Modulus
+
 ```k
     rule X modInt N => X
       requires 0 <=Int X
        andBool X  <Int N
       [simplification]
 ```
 
+`modInt` selects the least non-negative representative of a congruence class modulo `N`.
+
+```k
+    rule (X modInt M) modInt N => X modInt M
+      requires M >Int 0
+       andBool M <=Int N
+      [simplification]
+```
```

Comment:
Specific cae of the above. Add smt-lemma?
```
rule A ==Int X modInt M impliesBool 0 <=Int A andBool A <Int M => true [smt-lemma]
```

Reply:
Try adding smt-lemma and removing above lemma, still pass?
Investigate why z3 doesn't get this

Reply:
For thesis: document workflow

Reply:
Answer question: Why can't we just always posit all lemmas we know? Example: Associativity.
Reason: rewrite-based, infitine looping rewrites. Aiming for a normal form.

Reply:
(X + Y) mod N = (X mod N + Y mod N) mod N would solve a lot.

Reply:
Double check: all side conditions guarantee M and N are not 0